src/core_plugins/kibana/public/dashboard/grid.js

```diff
@@ -34,25 +68,27 @@ app.directive('dashboardGrid', function ($compile, Notifier) {
 
       // debounced layout function is safe to call as much as possible
       const safeLayout = _.debounce(layout, 200);
-
-      $scope.removePanelFromState = (panelIndex) => {
-        _.remove($scope.state.panels, function (panel) {
-          return panel.panelIndex === panelIndex;
-        });
-      };
-
       /**
-       * Removes the panel with the given index from the $scope.state.panels array. Does not
-       * remove the ui element from gridster - that is triggered by a watcher that is
-       * triggered on changes made to $scope.state.panels.
-       * @param panelIndex {number}
+       * Mapping of panelIndex to the angular element in the grid.
        */
-      $scope.getPanelByPanelIndex = (panelIndex) => {
-        return _.find($scope.state.panels, function (panel) {
-          return panel.panelIndex === panelIndex;
-        });
+      const panelElementMapping = {};
```

Comment:
You found some good cases where it was unnecessary, but I think we still need it.  In the watch function, it's used to grab the panels attached to each element, to compare the new panels with existing panels.  I don't think I can use the panelElement there.  Actually, I could probably do this by comparing the ids instead of the panels, but there is still one last place I need findPanelByPanelIndex.

It's this spot right here: 
```
 <dashboard-panel
                  remove="removePanel(${panel.panelIndex})"
                  panel="findPanelByPanelIndex(${panel.panelIndex}, panels)"
                  is-full-screen-mode="isFullScreenMode"
                  is-expanded="false"
                  get-vis-click-handler="getVisClickHandler(state)"
                  get-vis-brush-handler="getVisBrushHandler(state)"
                  save-state="saveState"
                  toggle-expand="toggleExpand(${panel.panelIndex})"
                  create-child-ui-state="createChildUiState">
```

I can't put the panel on the scope because then it would be overwritten every time, so I need the panels to be stored in an array on the scope.  I can't use the index in the array (e.g. `panels[index]`) because the panels change indexes after being deleted. I think that's the only way to link a the panel ui to it's  panel state, but open to other suggestions if you have any.